src/query_plan/query_planner.rs

```diff
@@ -106,26 +120,303 @@ impl Default for QueryPlannerDebugConfig {
+            });
+
+        let join_link = metadata.for_identity(&Identity::join_identity()).unwrap();
+        let join_field_directive =
+            join_link.directive_name_in_schema(&JOIN_FIELD_DIRECTIVE_NAME_IN_SPEC);
+
+        let is_interface_object =
+            |ty: &ExtendedType| ty.is_object() && ty.directives().has(&interface_object_directive);
+
+        let interface_types_with_interface_objects = supergraph
+            .schema
+            .get_types()
+            .filter_map(|position| match position {
+                TypeDefinitionPosition::Interface(interface_position) => Some(interface_position),
+                _ => None,
+            })
+            .filter(|position| {
+                query_graph.sources().any(|schema| {
+                    schema
+                        .schema()
+                        .types
+                        .get(&position.type_name)
+                        .is_some_and(is_interface_object)
+                })
+            })
+            .collect::<IndexSet<_>>();
+
+        let is_inconsistent = |position: AbstractTypeDefinitionPosition| {
+            let mut sources = query_graph.sources().filter_map(|subgraph| {
+                match subgraph.try_get_type(position.type_name().clone())? {
+                    TypeDefinitionPosition::Object(object) => Some(IndexSet::from([object])),
+                    TypeDefinitionPosition::Interface(interface) => Some(
+                        subgraph
+                            .referencers()
+                            .get_interface_type(&interface.type_name)
+                            .ok()?
+                            .object_types
+                            .clone(),
+                    ),
+                    TypeDefinitionPosition::Union(union_) => Some(
+                        union_
+                            .try_get(subgraph.schema())?
+                            .members
+                            .iter()
+                            .map(|member| ObjectTypeDefinitionPosition::new(member.name.clone()))
+                            .collect(),
+                    ),
+                    _ => None,
+                }
+            });
+
+            let Some(expected_runtimes) = sources.next() else {
+                return false;
+            };
+            sources.all(|runtimes| runtimes == expected_runtimes)
+        };
+
+        let abstract_types_with_inconsistent_runtime_types = supergraph
+            .schema
+            .get_types()
+            .filter_map(|position| AbstractTypeDefinitionPosition::try_from(position).ok())
+            .filter(|position| is_inconsistent(position.clone()))
+            .collect::<IndexSet<_>>();
+
+        // TODO(@goto-bus-stop): not sure if this is needed
+        // we can just do `.get().unwrap_or(false)` on a map with *some* labels,
+        // instead of preparing this map for all labels?
+        let _default_override_conditions = supergraph
```

Comment:
I think I agree. And `BuildQueryPlanOptions.overrideConditions` could be a set of strings (where presence means true) instead of a map of string to booleans